semanticdb/semanticdb3/semanticdb3.md

```diff
@@ -177,6 +177,251 @@ interspersed with a semicolon (`;`). The order of concatenation is unspecified.
 Must not be used outside `Synthetic.text` documents. Placeholder symbols are
 always equal to an asterisk (`*`).
 
+### Type
+
+```protobuf
+message Type {
+  enum Tag {
+    UNKNOWN_TAG = 0;
+    TYPE_REF = 1;
+    SINGLE_TYPE = 2;
+    THIS_TYPE = 3;
+    SUPER_TYPE = 4;
+    LITERAL_TYPE = 5;
+    COMPOUND_TYPE = 6;
+    ANNOTATED_TYPE = 7;
+    EXISTENTIAL_TYPE = 8;
+    TYPE_LAMBDA = 9;
+    CLASS_INFO_TYPE = 10;
+    METHOD_TYPE = 11;
+    BY_NAME_TYPE = 12;
+    REPEATED_TYPE = 13;
+    TYPE_TYPE = 14;
+  }
+  Tag tag = 1;
+  TypeRef typeRef = 2;
+  SingleType singleType = 3;
+  ThisType thisType = 4;
+  SuperType superType = 5;
+  LiteralType literalType = 6;
+  CompoundType compoundType = 7;
+  AnnotatedType annotatedType = 8;
+  ExistentialType existentialType = 9;
+  TypeLambda typeLambda = 10;
+  ClassInfoType classInfoType = 11;
+  MethodType methodType = 12;
+  ByNameType byNameType = 13;
+  RepeatedType repeatedType = 14;
+  TypeType typeType = 15;
+}
```

Comment:
This encoding is not for the faint-hearted. Can't say that I'm satisfied, but I don't currently see a way to make things better - probably because of my lack of experience with protobuf.

I experimented with [`oneof`](https://developers.google.com/protocol-buffers/docs/proto#oneof), but that was too much magic for my taste. It also didn't help that the official documentation has a dedicated section about backward compatibility issues associated with oneof fields - sounds like some scary stuff.